File: k8s/edge/edge.lua
--
--   Copyright 2018 EPAM Systems
--
--   Licensed under the Apache License, Version 2.0 (the "License");
--   you may not use this file except in compliance with the License.
--   You may obtain a copy of the License at
--
--       http://www.apache.org/licenses/LICENSE-2.0
--
--   Unless required by applicable law or agreed to in writing, software
--   distributed under the License is distributed on an "AS IS" BASIS,
--   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
--   See the License for the specific language governing permissions and
--   limitations under the License.
--
local os = require("os")
local uuid = require "resty.jit-uuid"
local cjson = require("cjson")

local _M = {}

function _M.get_config(name, default)
    local value = os.getenv(name)
    if value ~= Nil then
        return value
    else
        return default
    end
end


function _M.init_worker()
    uuid.seed()
end

function _M.generate_request_id()
    return uuid()
end

function _M.get_model_id_and_version_from_feedback_url()
    local m, err = ngx.re.match(ngx.var.uri, "/api/model/(.+)/(.+)/feedback")

    if err then
        ngx.exit(ngx.HTTP_BAD_REQUEST)
    end

    return m[1], m[2]
end

function _M.parse_feedback_data()
    local model_id, model_version = _M.get_model_id_and_version_from_feedback_url()
    local reqargs = require "resty.reqargs"
    local get, post, files = reqargs()
    local cjson = require "cjson"

    local data = {}
    local params = {}

    if post ~= nil and type(post) == "table" then
        for key, value in pairs(post) do
            params[key] = value
        end
    end

    if get ~= nil and type(get) == "table" then
        for key, value in pairs(get) do
            params[key] = value
        end
    end

    data["feedback_payload"] = params

    data["request_id"] = ngx.req.get_headers()["Request-ID"]

    data["model_id"] = model_id
    data["model_version"] = model_version

    return data
end

function _M.response_feedback(data)
    local response = {
        status = true,
        data = data
    }

    local json_encoded_response = cjson.encode(response)

    ngx.header["Content-Type"] = "application/json;charset=utf-8"
    ngx.say(json_encoded_response);
end

function _M.catch_model_api_response_chunk(model_id, model_version, content, eof)
    local http_request_headers = ngx.req.get_headers()
    local requestID = http_request_headers["Request-ID"]

    if eof and string.len(content) == 0 then
        return nil
    end

    local chunk_id = ngx.var.model_api_chunk_count

    local data = {
        request_id = requestID,
        response_chunk_id = chunk_id,
        response_content = content,
        model_id = model_id,
        model_version = model_version
    }

    ngx.var.model_api_chunk_count = chunk_id + 1

    return data
end

function _M.catch_model_api_call(model_id, model_version)
    local request_http_headers = ngx.req.get_headers()
    local request_id = request_http_headers["Request-ID"]
    local http_method = ngx.req.get_method()
    request_http_headers["authorization"] = nil

    local data = {
        request_id = request_id,
        request_http_method = http_method,
        request_http_headers = request_http_headers,
        request_host = ngx.var.HOST,
        request_uri = ngx.var.uri,
        response_duration = ngx.var.upstream_response_time,
        response_http_headers = ngx.resp.get_headers(),
        response_body_chunk_count = ngx.var.model_api_chunk_count,
        response_status = ngx.status,
        model_id = model_id,
        model_version = model_version
    }

    local content_type = ngx.var.content_type or ""
    if http_method == 'GET' then
        data['request_get_args'] = ngx.req.get_uri_args()
    elseif string.sub(content_type, 1, 33) == "application/x-www-form-urlencoded"
            and http_method == 'POST' then
        data["request_post_args"] = ngx.req.get_post_args()
    end

    return data
end

return _M

File: legion/sdk/legion/sdk/clients/edge.py
#
#    Copyright 2019 EPAM Systems
#
#    Licensed under the Apache License, Version 2.0 (the "License");
#    you may not use this file except in compliance with the License.
#    You may obtain a copy of the License at
#
#        http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS,
#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#    See the License for the specific language governing permissions and
#    limitations under the License.
#
"""
EDGE client
"""
import abc
import argparse
import logging
import time
import typing

import requests

from legion.sdk import config
from legion.sdk.containers.docker import build_docker_client, find_host_model_port
from legion.sdk import definitions
from legion.sdk.utils import normalize_name

DEFAULT_TIMEOUT = 10

DEFAULT_SLEEP_ITERATION = 10

LOGGER = logging.getLogger(__name__)


class EdgeClient(metaclass=abc.ABCMeta):
    """
    Base EDGE client
    """

    @abc.abstractmethod
    def invoke_model_api(self, model_id: str, model_version: str, payload: typing.Dict[str, typing.Any],
                         endpoint: str = 'default') -> typing.Any:
        """
        Perform inspect query on EDGE server

        :param model_id: model ID
        :param model_version: model version
        :param endpoint: name of endpoint
        :param payload: payload
        :return: json model response
        """
        pass

    @abc.abstractmethod
    def info(self, model_id: str, model_version: str) -> typing.Any:
        """
        Perform info query on EDGE server

        :param model_id: model ID
        :param model_version: model version
        :return: json model response
        """
        pass


class RemoteEdgeClient(EdgeClient):
    """
    EDGE client
    """

    def __init__(self, model_server_url: str, jwt: str, retries: int = 10) -> None:
        """
        Build RemoteEdgeClient

        :param model_server_url: edge url
        :param jwt: model jwt token
        """
        if not model_server_url:
            raise ValueError('Please, specify model server url')
        if not jwt:
            raise ValueError('Please, specify model jwt')

        self._model_server_url = model_server_url
        self._jwt = jwt
        self._retries = retries

    def _request(self, url: str, method: str = 'POST', payload: typing.Dict[str, typing.Any] = None) -> typing.Any:
        """
        Perform inspect query on EDGE server

        :param url: request url
        :param method: HTTP method
        :param payload: payload
        :return: json model response
        """
        headers = {'Authorization': f'Bearer {self._jwt}'}

        LOGGER.debug('Requesting {} with data = {} in POST mode'.format(url, payload))
        left_retries = self._retries if self._retries > 0 else 1
        raised_exception = None

        while left_retries > 0:
            try:
                LOGGER.debug('Requesting {}'.format(url))

                response = requests.request(method, url, data=payload, headers=headers, timeout=DEFAULT_TIMEOUT)

                if response.status_code == 401:
                    raise Exception(
                        'Wrong jwt model token. You can refresh it by using "legionctl generate-token" command')

                if not response.ok:
                    raise Exception(f'Returned wrong status code: {response.status_code}, text: {response.text}')

                return response.json()
            except (requests.exceptions.ConnectionError, requests.exceptions.Timeout) as exception:
                LOGGER.error('Failed to connect to {}: {}. Retrying'.format(url, exception))
                raised_exception = exception

                time.sleep(DEFAULT_SLEEP_ITERATION)

            left_retries -= 1

        raise Exception('Failed to connect to {}. No one retry left. Exception: {}'.format(
            url, raised_exception
        ))

    def invoke_model_api(self, model_id: str, model_version: str, payload: typing.Dict[str, typing.Any],
                         endpoint: str = 'default') -> typing.Any:
        """
        Perform inspect query on EDGE server

        :param model_id: model ID
        :param model_version: model version
        :param endpoint: name of endpoint
        :param payload: payload
        :return: json model response
        """
        url = f'{self._model_server_url}/api/model/{model_id}/{model_version}/invoke/{endpoint}'
        return self._request(url, payload=payload)

    def info(self, model_id: str, model_version: str) -> typing.Any:
        """
        Perform info query on EDGE server

        :param model_id: model ID
        :param model_version: model version
        :return: json model response
        """
        url = f'{self._model_server_url}/api/model/{model_id}/{model_version}/info'
        return self._request(url, method='GET')


class LocalEdgeClient(EdgeClient):

    def invoke_model_api(self, model_id: str, model_version: str, payload: typing.Dict[str, typing.Any],
                         endpoint: str = 'default') -> typing.Any:
        """
        Perform inspect query on EDGE server

        :param model_id: model ID
        :param model_version: model version
        :param endpoint: name of endpoint
        :param payload: payload
        :return: json model response
        """
        client = build_docker_client()
        containers = client.containers.list(filters={
            'label': [f'{definitions.LEGION_COMPONENT_LABEL}.model_id={model_id}',
                      f'{definitions.LEGION_COMPONENT_LABEL}.model_version={model_version}']
        })

        if not containers:
            raise ValueError(
                f'Can not find container with {model_id} model id and {model_version} model version labels')

        if len(containers) > 1:
            raise ValueError(
                f'Find multiple containers with {model_id} model id and {model_version} model version labels')

        model_port = find_host_model_port(containers[0])

        url = f'http://localhost:{model_port}/api/model/{model_id}/{model_version}/invoke/{endpoint}'
        response = requests.post(
            url,
            data=payload
        )

        if not response.ok:
            raise Exception(f'Returned wrong status code: {response.status_code}, text: {response.text}')

        return response.json()

    def info(self, model_id: str, model_version: str) -> typing.Any:
        """
        Stub method
        :param model_id: model id
        :param model_version: model version
        """
        pass


def model_config_prefix(model_id: str, model_version: str) -> str:
    """
    Build model prefix for the cli config
    :param model_id: model id
    :param model_version: model version
    :return: config prefix
    """
    return normalize_name(f'{model_id}-{model_version}')


def build_client(args: argparse.Namespace = None) -> EdgeClient:
    """
    Build EDGE client from from ENV and from command line arguments

    :param args: (optional) command arguments with .namespace
    :return: EDGE client
    """
    host, token = None, None

    if args:
        if 'local' in args and args.local:
            return LocalEdgeClient()

        if args.model_server_url:
            host = args.model_server_url

        if args.jwt:
            token = args.jwt

    host = host or config.MODEL_SERVER_URL
    token = token or config.get_config_file_variable(model_config_prefix(args.model_id, args.model_version),
                                                     section=config.MODEL_JWT_TOKEN_SECTION)

    return RemoteEdgeClient(host, token)
